### ISSUE TITLE
Initializing respective weights by different method for TorchLayer

### ISSUE THREAD

**Issue by shukob** [PennyLaneAI/pennylane#898]
#### Feature request

In `TorchLayer`, we cannot specify per weight initialization method. Furthermore, the `init_method` only accepts a function that takes one argument: `shape`. Because of this, it is not possible to initialize, for example,  multiple entangling layers in one qnode with different initialization methods. 

* *Possible workaround:* 

We can specify a `dict` for init_method, each key is a `weight` name and each value is the weight itself or a callable to produce initial weight. Or we can specify a callable which takes `weight` name so that we can conditionally return the respective initial weight or callable.

* *System information:* (post the output of `import pennylane as qml; qml.about()`)
Name: PennyLane
Version: 0.12.0
Summary: PennyLane is a Python quantum machine learning library by Xanadu Inc.
Home-page: https://github.com/XanaduAI/pennylane
Author: None
Author-email: None
License: Apache License 2.0
Location: /home/shukob/anaconda3/envs/SkipGram/lib/python3.7/site-packages
Requires: appdirs, numpy, semantic-version, scipy, toml, autograd, networkx
Required-by: pennylane-qulacs, PennyLane-qiskit
Platform info:           Linux-5.4.0-52-generic-x86_64-with-debian-bullseye-sid
Python version:          3.7.9
Numpy version:           1.19.2
Scipy version:           1.5.2
Installed devices:
- default.gaussian (PennyLane-0.12.0)
- default.mixed (PennyLane-0.12.0)
- default.qubit (PennyLane-0.12.0)
- default.qubit.autograd (PennyLane-0.12.0)
- default.qubit.tf (PennyLane-0.12.0)
- default.tensor (PennyLane-0.12.0)
- default.tensor.tf (PennyLane-0.12.0)
- qulacs.simulator (pennylane-qulacs-0.12.0)
- qiskit.aer (PennyLane-qiskit-0.12.0)
- qiskit.basicaer (PennyLane-qiskit-0.12.0)
- qiskit.ibmq (PennyLane-qiskit-0.12.0)

**Comment by trbromley** [comment]
Thanks @shukob for this feature request!

If anyone is interested to work on this feature, check out our contribution guide [here](https://discuss.pennylane.ai/t/contributing-to-pennylane/15).

**Comment by QRiddler** [comment]
Hi! Can I work on this issue?

**Comment by antalszava** [comment]
Hi @QRiddler, sure. :slightly_smiling_face: Feel free to ask any questions or open a work in progress PR for it!

**Comment by antalszava** [comment]
Hi @anonymousr007, this issue is related to extending the [`TorchLayer` class in PennyLane](https://pennylane.readthedocs.io/en/stable/code/api/pennylane.qnn.TorchLayer.html?highlight=torchlayer#pennylane.qnn.TorchLayer).

When creating an instance of this class, we have the ability to pass an `init_method` argument :

```markdown
init_method (callable): a `torch.nn.init <https://pytorch.org/docs/stable/nn.init.html>`__ function for initializing the QNode weights. If not specified, weights are randomly initialized using the uniform distribution over :math:`[0, 2 \pi]`.
```
As the issue description suggests, we'd aim to extend the logic using the `init_method` argument such that:
* An initialization method can be specified for each weight if desired;
* Initialization methods can be specified for separate layers within templates (e.g., multiple entangling layers in one qnode with different initialization methods).

**Comment by amitjansc** [comment]
Hello! I will be working on this issue.

I have a small question: In your [contribution guide](https://discuss.pennylane.ai/t/contributing-to-pennylane/15) you ask that the code conforms to the PEP8 coding standard, which includes a maximum line length of 79 characters. However this is not taken into account in some parts of the code. What limit should I use? 79, 99 or 119?

**Comment by josh146** [comment]
> In your [contribution guide](https://discuss.pennylane.ai/t/contributing-to-pennylane/15) you ask that the code conforms to the PEP8 coding standard, which includes a maximum line length of 79 characters. However this is not taken into account in some parts of the code. What limit should I use? 79, 99 or 119?

Hey @amitjansc! We use a line length of 100, which you can autoformat using `black -l 100 <file>` :)

Regarding your second question, I am not too sure - maybe @antalszava knows?

**Comment by amitjansc** [comment]
In PR https://github.com/PennyLaneAI/pennylane/pull/2678 I changed the following `if` condition such that size is casted to an emtpy tuple only when it is 0 or 1. Consequently an error is raised when defining a negative weight shape.
https://github.com/PennyLaneAI/pennylane/blob/3344c77e39354df70b7ca989a65f5fe982b63479/pennylane/qnn/torch.py#L211

However, I still don't understand why do you use a scalar Tensor for `size = 1`. Shouldn't it be a 1-dim Tensor with size 1?

**Comment by trbromley** [comment]
Hi @amitjansc!

> In PR https://github.com/PennyLaneAI/pennylane/pull/2678 I changed the following if condition such that size is casted to an emtpy tuple only when it is 0 or 1.

Thanks, that sounds like a good idea! You could potentially exclude the `0` value too.

> However, I still don't understand why do you use a scalar Tensor for size = 1. Shouldn't it be a 1-dim Tensor with size 1?

I worked on this feature a few years back, so apologies if my memory is a little unclear. I believe the intention was to support the case where a user wants to work with a scalar tensor, i.e., so
```python
weight_shapes = {"theta": 1}
```
would support using a single `theta` angle in the circuit.

Of course, we could perhaps more accurately require `weight_shapes = {"theta": ()}`, but I think we decided that `{"theta": 1}` was more to the expectation of a user. In the eventuality that a user wishes to work with a rank-1 tensor that happens to be 1-dimensional, they could still do `{"theta": [1]}`.
